### PR TITLE
Implement RemoveServiceAccount in AWS warehouse

### PR DESCRIPTION
**Pull request by mbarkley**
Implementation and unit tests for the `RemoveServiceAccount` method on the AWS warehouse.

### CODE REVIEW

File: lib/aws/warehouse.go [hunk 1]
```diff
@@ -198,17 +202,85 @@ func (wh *AccountWarehouse) GetServiceAccounts(ctx context.Context, _ string) (<
...
+		_, err = wh.apiClient.DeleteLoginProfile(&iam.DeleteLoginProfileInput{UserName: aws.String(userName)})
+		if err != nil {
+			return fmt.Errorf("unable to delete AWS user %s login profile: %v", userName, err)
+		}
+	} else if aerr, ok := err.(awserr.Error); !ok || aerr.Code() != iam.ErrCodeNoSuchEntityException {
+		return fmt.Errorf("error looking up login profile while attempting to delete AWS user %s: %v", userName, aerr)
+	}
+
+	// delete access keys
+	listKeysOutput, err := wh.apiClient.ListAccessKeys(&iam.ListAccessKeysInput{UserName: aws.String(userName)})
```
Comment (chaopeng) [inline lib/aws/warehouse.go:220]: Maybe split this part to a new func listUserAccessKeys or deleteUserAccessKeys same below for inline policies.

Also we maybe we should run in order:

- delete AccessKeys
- delete inline policies
- delete DeleteLoginProfile
- delete user

to make this endpoints retriable.

File: lib/aws/warehouse.go [hunk 2]
```diff
@@ -198,17 +202,85 @@ func (wh *AccountWarehouse) GetServiceAccounts(ctx context.Context, _ string) (<
...
+	if err == nil {
+		_, err = wh.apiClient.DeleteLoginProfile(&iam.DeleteLoginProfileInput{UserName: aws.String(userName)})
+		if err != nil {
+			return fmt.Errorf("unable to delete AWS user %s login profile: %v", userName, err)
+		}
+	} else if aerr, ok := err.(awserr.Error); !ok || aerr.Code() != iam.ErrCodeNoSuchEntityException {
+		return fmt.Errorf("error looking up login profile while attempting to delete AWS user %s: %v", userName, aerr)
+	}
+
+	// delete access keys
```
Comment (chaopeng) [inline lib/aws/warehouse.go:219]: maybe add a var `marker := ""` to make the first call same as following calls

File: lib/aws/client.go [hunk 3]
```diff
@@ -301,6 +336,42 @@ func (m *MockAwsClient) PutUserPolicy(input *iam.PutUserPolicyInput) (*iam.PutUs
 	return &iam.PutUserPolicyOutput{}, nil
 }
 
+// ListUserPolicies ...
+func (m *MockAwsClient) ListUserPolicies(input *iam.ListUserPoliciesInput) (*iam.ListUserPoliciesOutput, error) {
+	if _, err := m.GetUser(&iam.GetUserInput{UserName: input.UserName}); err != nil {
+		return nil, err
+	}
+	policyNames := make([]*string, 0)
```
Comment (chaopeng) [inline lib/aws/client.go:344]: nit: var policyNames []*string

File: lib/aws/client.go [hunk 4]
```diff
@@ -301,6 +336,42 @@ func (m *MockAwsClient) PutUserPolicy(input *iam.PutUserPolicyInput) (*iam.PutUs
...
+	}, nil
+}
+
+// DeleteUserPolicy ...
+func (m *MockAwsClient) DeleteUserPolicy(input *iam.DeleteUserPolicyInput) (*iam.DeleteUserPolicyOutput, error) {
+	if _, err := m.GetUser(&iam.GetUserInput{UserName: input.UserName}); err != nil {
+		return nil, err
+	}
+
+	newPolicies := make([]*iam.PutUserPolicyInput, 0)
```
Comment (chaopeng) [inline lib/aws/client.go:364]: same here

File: lib/aws/client.go [hunk 5]
```diff
@@ -343,6 +414,21 @@ func (m *MockAwsClient) CreateUser(input *iam.CreateUserInput) (*iam.CreateUserO
 	}, nil
 }
 
+// DeleteUser ...
+func (m *MockAwsClient) DeleteUser(input *iam.DeleteUserInput) (*iam.DeleteUserOutput, error) {
+	for i, user := range m.Users {
+		if *input.UserName == *user.UserName {
+			newUsers := make([]*iam.User, len(m.Users)-1)
+			copy(newUsers, m.Users[0:i])
```
Comment (chaopeng) [inline lib/aws/client.go:422]: maybe more readable:

```go
var newUsers []*iam.User
var found bool
for _, user := range m.Users {
	if *input.UserName == *user.UserName {
		found = true
		continue
	}
	newUsers = append(newUsers, user)
}
if !found {
	...
}
m.Users = newUsers
...
```

File: lib/aws/warehouse.go (hunk 1 above)
Comment (mbarkley) [reply lib/aws/warehouse.go:220]: @chaopeng Working on your other suggestions, but regarding the order of operations, this is the order recommended by AWS CLI documentation. The login profile is orthogonal to access keys and policies (it is specifically for username/password access to the AWS web console) so it should still be possible to retry removing an account if any one of those operations fails.

**Comment by mbarkley** [comment]
@chaopeng updated with most of the changes you suggested. Please see my comment in-line about the order of operations in `RemoveServiceAccount` and let me know if that sounds good to you or if there is something you were saying that I missed.

**Comment by chaopeng** [comment]
lgtm, @cdvoisin ptal.

File: lib/aws/client.go [hunk 6]
```diff
@@ -343,6 +414,25 @@ func (m *MockAwsClient) CreateUser(input *iam.CreateUserInput) (*iam.CreateUserO
 	}, nil
 }
 
+// DeleteUser ...
+func (m *MockAwsClient) DeleteUser(input *iam.DeleteUserInput) (*iam.DeleteUserOutput, error) {
+	var newUsers[]*iam.User
```
Comment (cdvoisin) [inline lib/aws/client.go:419]: add a space between variable name and definition: "var newUsers []*iam.User"

Comment (mbarkley) [reply lib/aws/client.go]: Fixed.

**Pull request closed without merging**